src/sage/combinat/designs/covering_array.py

```diff
@@ -50,3 +54,190 @@
+        sage: is_covering_array(E,parameters=True)
+        (True, (11, 2, 4, 3))
+
+    """
+    oldk = len(array[0])
+
+    if oldk == k:
+        return array
+
+    elif oldk < k:
+        raise ValueError("array only has {} columns".format(oldk))
+
+    else:
+        result = []
+        for row in array:
+            result.append(row[:k])
+        return result
+
+
+def Kleitman_Spencer_Katona(N):
+    r"""
+    Return a `CA(N; 2, k, 2)` where `k = \binom {N-1}{\lceil \frac{N}{2} \rceil}`.
+
+    INPUT:
+
+    - ``N`` -- the number of rows in the array, must be an integer greater
+      than 3 since any smaller would not produce enough columns for a
+      strength 2 array.
+
+    This construction is referenced in [Colb2004]_ from [KS1973]_ and [Kat1973]_
+
+    **Construction**
+
+    Take all distinct binary `N`-tuples of weight `\frac{N}{2}` that have a 0
+    in the first position and place them as columns in an array.
+
+    EXAMPLES::
+
+        sage: from sage.combinat.designs.covering_array import Kleitman_Spencer_Katona
+        sage: from sage.combinat.designs.designs_pyx import is_covering_array
+        sage: C = Kleitman_Spencer_Katona(2)
+        Traceback (most recent call last):
+        ...
+        ValueError: N must be greater than 3
+        sage: C = Kleitman_Spencer_Katona(5)
+        sage: is_covering_array(C,parameters=True)
+        (True, (5, 2, 4, 2))
+
+    """
+    from math import ceil
+    from itertools import combinations
+
+    if N < 4:
+        raise ValueError("N must be greater than 3")
+
+    result = []
+    for p in combinations(range(N-1), ceil(N/2)):
+        S = [0]*N
+        for i in p:
+            S[i] = 1
+        result.append(S)
+    return(list(map(list, zip(*result))))
+
+
+def column_Kleitman_Spencer_Katona(k):
+    r"""
+    Return a covering array with `k` columns using the Kleitman Spencer Katona
+    method.
+
+    See :func:`~sage.combinat.designs.covering_array.Kleitman_Spencer_Katona`
+
+    INPUT:
+
+    - ``k`` -- the number of columns in the array, must be an integer
+      greater than 3 since any smaller is a trivial array for strength 2.
+
+    EXAMPLES::
+
+        sage: from sage.combinat.designs.designs_pyx import is_covering_array
+        sage: from sage.combinat.designs.covering_array import column_Kleitman_Spencer_Katona
+        sage: C = column_Kleitman_Spencer_Katona(20)
+        sage: is_covering_array(C,parameters=True)
+        (True, (8, 2, 20, 2))
+        sage: column_Kleitman_Spencer_Katona(25000)
+        Traceback (most recent call last):
+        ...
+        AssertionError: not implemented for k > 24310
+
+    """
+    assert k <= 24310, "not implemented for k > 24310"
+
+    kdict = {3:4,4:5,10:6,15:7,35:8,56:9,126:10,210:11,462:12,792:13,
+            1716:14,3003:15,6435:16,11440:17,24310:18}
+
+    for ki in kdict:
+        if k <= ki:
+            N = kdict[ki]
+            break
+    return truncate_columns(Kleitman_Spencer_Katona(N), k)
```

Comment:
The `kdict` approach may not work: Python does not guarantee that dictionaries maintain their order. If it places the 24310 at the beginning, for example, you will always get 18.

I would implement this as a list of tuple:
```
kdict = [(3, 4), (4, 5), (10, 6), (15, 7), (35, 8), (56, 9), (126, 10), (210, 11), (462, 12), (792, 13), (1716, 14), (3003, 15), (6435, 16), (11440, 17), (24310, 18)]
```

I also wouldn't hard-copy the maximum value of `k` in the code like you do above. If you can ever extend `kdict`, then you'll have to change it in two places. I would do (after the above definition):

```
assert k <= kdict[-1][0]
```

Then the rest of the logic turns into:

```
for ki, N in kdict:
    if k <= ki:
        return truncate_columns(Kleitman_Spencer_Katona(N), k)
```

Depending on the version of Python you're using, all of this can probably be replaced by the `bisect` library, using, I believe, the `bisect_left` method.

Reply:
The first suggestion is much appreciated. Also bisect_left was very interesting to learn about but I am not sure what is faster when the lists are so small like this, even if it is expanded it will not be expanded much.

Reply:
> Python does not guarantee that dictionaries maintain their order.

Not true any more since Python 3.7. Dictionary maintain insertion order

Reply:
Yes, you are correct, @aadwyer: maps are now ordered.

Reply:
Because there are only a small number of entries, the NotImplementedError is raised fast and it will often have to read each entry in either case, I will keep this as a list despite the fact dictionaries could work now.